compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1785,28 +1785,42 @@ trait Applications extends Compatibility {
       candidates.flatMap(cloneCandidate)
     }
 
+    def resultIsMethod(tp: Type): Boolean = tp.widen.stripPoly match
+      case tp: MethodType => tp.resultType.isInstanceOf[MethodType]
+      case _ => false
+
     val found = narrowMostSpecific(candidates)
     if (found.length <= 1) found
-    else pt match {
-      case pt @ FunProto(_, resType: FunProto) =>
-        // try to narrow further with snd argument list
-        val advanced = advanceCandidates(pt.typedArgs().tpes)
-        resolveOverloaded(advanced.map(_._1), resType, Nil) // resolve with candidates where first params are stripped
-          .map(advanced.toMap) // map surviving result(s) back to original candidates
-      case _ =>
-        val noDefaults = alts.filter(!_.symbol.hasDefaultParams)
-        val noDefaultsCount = noDefaults.length
-        if (noDefaultsCount == 1)
-          noDefaults // return unique alternative without default parameters if it exists
-        else if (noDefaultsCount > 1 && noDefaultsCount < alts.length)
-          resolveOverloaded(noDefaults, pt, targs) // try again, dropping defult arguments
-        else {
-          val deepPt = pt.deepenProto
-          if (deepPt ne pt) resolveOverloaded(alts, deepPt, targs)
-          else candidates
-        }
-    }
-  }
+    else
+      val deepPt = pt.deepenProto
+      deepPt match
+        case pt @ FunProto(_, resType: FunProto) =>
+          // try to narrow further with snd argument list
+          val advanced = advanceCandidates(pt.typedArgs().tpes)
+          resolveOverloaded(advanced.map(_._1), resType, Nil) // resolve with candidates where first params are stripped
+            .map(advanced.toMap) // map surviving result(s) back to original candidates
+        case _ =>
+          // prefer alternatives that need no eta expansion
+          val noCurried = alts.filter(!resultIsMethod(_))
+          val noCurriedCount = noCurried.length
+          if noCurried.length == 1 then
```

Comment:
As pointed out by @michelou ,  we can just use `noCurriedCount ` instead of `noCurried.length `.